File: packages/sdk/src/integrations/mcp.ts
/**
 * AgentGuard MCP Integration — Node SDK
 *
 * Provides MCP (Model Context Protocol) enforcement for Node.js MCP servers.
 *
 * Usage:
 * ```typescript
 * import { AgentGuard } from 'agentguard';
 * import { createMcpProxy, wrapMcpServer } from 'agentguard/mcp';
 *
 * // Option 1: Wrap an existing MCP server (in-process)
 * const server = new MyMcpServer();
 * const protected = agentguard.mcp.wrapServer(server, { agentId: 'my-agent' });
 *
 * // Option 2: Create a proxy MCP server
 * const proxy = agentguard.mcp.createProxy({
 *   upstreamUrl: 'http://localhost:3000',
 *   agentId: 'my-agent',
 *   port: 3100, // AGENTGUARD_MCP_PROXY_PORT
 * });
 * await proxy.start();
 * ```
 */

// ─── Types ────────────────────────────────────────────────────────────────────

export interface McpProxyOptions {
  /** Port to listen on (default: 3100, or AGENTGUARD_MCP_PROXY_PORT env var) */
  port?: number;
  /** Upstream MCP server URL */
  upstreamUrl: string;
  /** AgentGuard agent ID */
  agentId: string;
  /** AgentGuard API base URL (default: https://api.agentguard.tech) */
  agentguardUrl?: string;
  /** AgentGuard API key (default: AGENTGUARD_API_KEY env var) */
  apiKey?: string;
  /** Session ID prefix */
  sessionPrefix?: string;
  /** Upstream request timeout in ms (default: 30000) */
  upstreamTimeoutMs?: number;
  /** Block on policy eval errors (default: true) */
  strict?: boolean;
}

export interface McpServerWrapOptions {
  /** AgentGuard agent ID */
  agentId: string;
  /** Session ID (or prefix for generated IDs) */
  sessionId?: string;
  /** AgentGuard API base URL */
  agentguardUrl?: string;
  /** AgentGuard API key */
  apiKey?: string;
  /** Block on policy eval errors */
  strict?: boolean;
}

export interface McpInterceptResult {
  allowed: boolean;
  decision: 'allow' | 'block' | 'hitl' | 'monitor';
  reason: string;
  riskScore: number;
  matchedRuleId?: string | null;
  gateId?: string | null;
  gateTimeoutSec?: number | null;
  evaluationMs: number;
}

export interface McpProxyServer {
  /** Start the proxy server */
  start(): Promise<void>;
  /** Stop the proxy server */
  stop(): void;
  /** The port the server is listening on */
  port: number;
}

// ─── McpClient — calls AgentGuard API for policy evaluation ──────────────────

class McpClient {
  private readonly baseUrl: string;
  private readonly apiKey: string;
  private readonly agentId: string;

  constructor(options: { baseUrl: string; apiKey: string; agentId: string }) {
    this.baseUrl = options.baseUrl;
    this.apiKey = options.apiKey;
    this.agentId = options.agentId;
  }

  async intercept(
    toolName: string,
    args: Record<string, unknown>,
    sessionId: string,
  ): Promise<McpInterceptResult> {
    const response = await fetch(`${this.baseUrl}/v1/mcp/intercept`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${this.apiKey}`,
      },
      body: JSON.stringify({
        request: {
          method: 'tools/call',
          id: `sdk-${Date.now()}`,
          params: { name: toolName, arguments: args },
        },
        identity: {
          agentId: this.agentId,
          sessionId,
        },
      }),
    });

    const data = await response.json() as McpInterceptResult;
    return data;
  }
}

// ─── createProxy — creates a standalone MCP proxy server ─────────────────────

/**
 * Create an MCP proxy server that intercepts tool calls.
 *
 * The proxy sits transparently between the AI agent and the MCP server:
 *   Agent → AgentGuard MCP Proxy (port 3100) → Upstream MCP Server
 *
 * @param options - Proxy configuration
 * @returns McpProxyServer with start/stop methods
 *
 * @example
 * ```typescript
 * const proxy = createMcpProxy({
 *   upstreamUrl: 'http://localhost:3000',
 *   agentId: 'my-agent',
 *   port: 3100,
 * });
 * await proxy.start();
 * console.log('MCP proxy running on port 3100');
 * ```
 */
export function createMcpProxy(options: McpProxyOptions): McpProxyServer {
  const port = options.port ?? parseInt(process.env['AGENTGUARD_MCP_PROXY_PORT'] ?? '3100', 10);
  const apiKey = options.apiKey ?? process.env['AGENTGUARD_API_KEY'] ?? '';
  const agentguardUrl = options.agentguardUrl ?? process.env['AGENTGUARD_API_URL'] ?? 'https://api.agentguard.tech';
  const client = new McpClient({ baseUrl: agentguardUrl, apiKey, agentId: options.agentId });

  let server: { close: () => void } | null = null;
  const upstreamTimeoutMs = options.upstreamTimeoutMs ?? 30000;

  const start = async (): Promise<void> => {
    // Dynamically import Hono to avoid hard dependency when proxy mode isn't used
    const { Hono } = await import('hono');
    const { serve } = await import('@hono/node-server');

    const app = new Hono();

    // Intercept POST requests (JSON-RPC)
    app.post('/', async (c) => {
      const body = await c.req.json<{ method?: string; id?: string | number; params?: { name?: string; arguments?: Record<string, unknown> } }>().catch(() => null);

      if (!body) {
        return c.json({ error: { code: -32700, message: 'Parse error' } }, 400);
      }

      // Only intercept tools/call
      if (body.method !== 'tools/call') {
        return c.json(await forwardRequest(options.upstreamUrl, body, upstreamTimeoutMs));
      }

      const toolName = body.params?.name ?? '';
      const toolArgs = body.params?.arguments ?? {};
      const sessionId = c.req.header('x-session-id') ??
        `${options.sessionPrefix ?? 'mcp'}_${Date.now()}`;

      // Evaluate against policy
      let result: McpInterceptResult;
      try {
        result = await client.intercept(toolName, toolArgs, sessionId);
      } catch (err) {
        if (options.strict !== false) {
          return c.json({
            id: body.id,
            error: { code: -32603, message: `AgentGuard policy evaluation failed: ${String(err)}` },
          }, 503);
        }
        // Permissive mode — allow on error
        return c.json(await forwardRequest(options.upstreamUrl, body, upstreamTimeoutMs));
      }

      // Handle blocked
      if (!result.allowed && result.decision === 'block') {
        return c.json({
          id: body.id,
          error: {
            code: -32603,
            message: `Tool call blocked by AgentGuard: ${result.reason}`,
            data: { decision: result.decision, riskScore: result.riskScore },
          },
        }, 403);
      }

      // Handle HITL pending
      if (result.decision === 'hitl') {
        return c.json({
          id: body.id,
          result: {
            content: [{
              type: 'text',
              text: JSON.stringify({
                status: 'hitl_pending',
                gateId: result.gateId,
                gateTimeoutSec: result.gateTimeoutSec,
                message: 'Tool call pending human approval',
              }),
            }],
          },
        }, 202);
      }

      // Allowed — forward to upstream
      const upstream = await forwardRequest(options.upstreamUrl, body, upstreamTimeoutMs);
      return c.json(upstream);
    });

    app.get('/health', (c) => c.json({ status: 'ok', agentId: options.agentId, port }));

    return new Promise<void>((resolve) => {
      const s = serve({ fetch: app.fetch, port });
      server = s as unknown as { close: () => void };
      console.log(`[AgentGuard MCP Proxy] Listening on :${port} → ${options.upstreamUrl}`);
      resolve();
    });
  };

  const stop = (): void => {
    if (server?.close) server.close();
  };

  return { start, stop, port };
}

// ─── wrapMcpServer — wraps an MCP server object in-process ───────────────────

/**
 * Wrap an existing MCP server with AgentGuard policy enforcement.
 *
 * This wraps the server's tool handler to intercept every tool call
 * before execution, evaluating it against the agent's policy.
 *
 * @param server - An MCP server instance with a `handleToolCall` method
 * @param options - Wrap options
 * @returns A wrapped server with the same interface
 *
 * @example
 * ```typescript
 * import { Server } from '@modelcontextprotocol/sdk/server/index.js';
 * import { wrapMcpServer } from 'agentguard/mcp';
 *
 * const server = new Server({ name: 'my-server', version: '1.0' });
 * const protected = wrapMcpServer(server, { agentId: 'my-agent' });
 *
 * // Use protected instead of server
 * protected.connect(transport);
 * ```
 */
export function wrapMcpServer<T extends Record<string, unknown>>(
  server: T,
  options: McpServerWrapOptions,
): T {
  const apiKey = options.apiKey ?? process.env['AGENTGUARD_API_KEY'] ?? '';
  const agentguardUrl = options.agentguardUrl ?? process.env['AGENTGUARD_API_URL'] ?? 'https://api.agentguard.tech';
  const client = new McpClient({ baseUrl: agentguardUrl, apiKey, agentId: options.agentId });

  // Create a proxy that intercepts tool call handler methods
  return new Proxy(server, {
    get(target, prop) {
      const value = (target as Record<string | symbol, unknown>)[prop as string];

      // Wrap setRequestHandler to intercept tools/call registration
      if (prop === 'setRequestHandler' && typeof value === 'function') {
        return function (schema: unknown, handler: (...args: unknown[]) => Promise<unknown>) {
          // Check if this is a tools/call handler
          const schemaAny = schema as { method?: string };
          if (schemaAny?.method === 'tools/call') {
            const wrappedHandler = async (...args: unknown[]) => {
              const requestArg = args[0] as { params?: { name?: string; arguments?: Record<string, unknown> } };
              const toolName = requestArg?.params?.name ?? 'unknown';
              const toolArgs = requestArg?.params?.arguments ?? {};
              const sessionId = options.sessionId ?? `server-${Date.now()}`;

              // Evaluate policy
              let result: McpInterceptResult;
              try {
                result = await client.intercept(toolName, toolArgs, sessionId);
              } catch (err) {
                if (options.strict !== false) {
                  throw new Error(`AgentGuard policy evaluation failed: ${String(err)}`);
                }
                // Allow on error in permissive mode
                return handler(...args);
              }

              if (!result.allowed && result.decision === 'block') {
                throw new Error(`AgentGuard: Tool "${toolName}" blocked — ${result.reason}`);
              }

              if (result.decision === 'hitl') {
                return {
                  content: [{
                    type: 'text',
                    text: JSON.stringify({
                      status: 'hitl_pending',
                      gateId: result.gateId,
                      message: `Tool "${toolName}" requires human approval`,
                    }),
                  }],
                };
              }

              return handler(...args);
            };

            return (value as (...a: unknown[]) => unknown).call(target, schema, wrappedHandler);
          }

          return (value as (...a: unknown[]) => unknown).call(target, schema, handler);
        };
      }

      if (typeof value === 'function') {
        return (value as (...args: unknown[]) => unknown).bind(target);
      }

      return value;
    },
  }) as T;
}

// ─── Helpers ──────────────────────────────────────────────────────────────────

async function forwardRequest(
  upstreamUrl: string,
  body: unknown,
  timeoutMs: number,
): Promise<unknown> {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), timeoutMs);
  try {
    const response = await fetch(upstreamUrl, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(body),
      signal: controller.signal,
    });
    return response.json();
  } finally {
    clearTimeout(timer);
  }
}

File: packages/python/agentguard/integrations/mcp.py
"""AgentGuard — MCP Integration (Python).

Provides MCP (Model Context Protocol) runtime policy enforcement.

Usage::

    # Option 1: Wrap an existing MCP server (in-process interception)
    from mcp.server import Server
    from agentguard.integrations.mcp import wrap_server

    server = Server("my-server")
    protected = wrap_server(
        server,
        api_key="ag_...",
        agent_id="my-agent",
    )

    # Use `protected` instead of `server` — all tool calls go through AgentGuard
    protected.run(transport="stdio")


    # Option 2: Intercept individual tool calls manually
    from agentguard.integrations.mcp import McpGuard

    guard = McpGuard(api_key="ag_...", agent_id="my-agent")

    async def my_tool_handler(tool_name: str, arguments: dict):
        decision = await guard.intercept(tool_name, arguments, session_id="sess-001")
        if not decision.allowed:
            raise PermissionError(f"Tool blocked: {decision.reason}")
        return await execute_tool(tool_name, arguments)


    # Option 3: Use as an async context manager decorator
    @guard.enforce(session_id="sess-001")
    async def my_tool(path: str) -> str:
        return open(path).read()
"""
from __future__ import annotations

import asyncio
import json
import os
import time
import urllib.request
from dataclasses import dataclass, field
from functools import wraps
from typing import Any, Callable, Dict, List, Optional


# ─── Data Classes ─────────────────────────────────────────────────────────────

@dataclass
class McpDecision:
    """Result of an AgentGuard MCP policy evaluation."""
    allowed: bool
    decision: str  # 'allow' | 'block' | 'hitl' | 'monitor'
    reason: str
    risk_score: int
    matched_rule_id: Optional[str] = None
    gate_id: Optional[str] = None
    gate_timeout_sec: Optional[int] = None
    evaluation_ms: float = 0.0

    @property
    def is_hitl(self) -> bool:
        return self.decision == "hitl"

    @property
    def is_blocked(self) -> bool:
        return self.decision == "block"

    @property
    def is_monitor(self) -> bool:
        return self.decision == "monitor"


@dataclass
class McpGuardConfig:
    """Configuration for the MCP guard."""
    api_key: str
    agent_id: str
    base_url: str = "https://api.agentguard.tech"
    session_id: Optional[str] = None
    strict: bool = True
    timeout_sec: float = 10.0


# ─── McpGuard ─────────────────────────────────────────────────────────────────

class McpGuard:
    """
    AgentGuard MCP Policy Enforcer.

    Evaluates MCP tool calls against the agent's active policy via the
    AgentGuard API. Supports blocking, monitoring, and HITL gating.

    Example::

        guard = McpGuard(api_key="ag_...", agent_id="my-agent")

        # Synchronous check
        decision = guard.intercept_sync("filesystem_write", {"path": "/etc/passwd"})
        if not decision.allowed:
            raise PermissionError(decision.reason)

        # Async check
        decision = await guard.intercept("web_request", {"url": "https://example.com"})
    """

    def __init__(
        self,
        api_key: str,
        agent_id: str,
        base_url: str = "https://api.agentguard.tech",
        session_id: Optional[str] = None,
        strict: bool = True,
        timeout_sec: float = 10.0,
    ) -> None:
        self._config = McpGuardConfig(
            api_key=api_key or os.environ.get("AGENTGUARD_API_KEY", ""),
            agent_id=agent_id or os.environ.get("AGENTGUARD_AGENT_ID", ""),
            base_url=base_url or os.environ.get("AGENTGUARD_API_URL", "https://api.agentguard.tech"),
            session_id=session_id,
            strict=strict,
            timeout_sec=timeout_sec,
        )
        self._session_id = session_id or f"mcp_{int(time.time())}"

    def intercept_sync(
        self,
        tool_name: str,
        arguments: Dict[str, Any],
        session_id: Optional[str] = None,
    ) -> McpDecision:
        """
        Synchronously evaluate an MCP tool call against the agent's policy.

        :param tool_name: The name of the MCP tool being called.
        :param arguments: The arguments passed to the tool.
        :param session_id: Session ID for audit continuity (optional).
        :returns: McpDecision with the policy result.
        """
        sid = session_id or self._session_id
        start = time.monotonic()

        payload = json.dumps({
            "request": {
                "method": "tools/call",
                "id": f"py-sdk-{int(time.time() * 1000)}",
                "params": {
                    "name": tool_name,
                    "arguments": arguments,
                },
            },
            "identity": {
                "agentId": self._config.agent_id,
                "sessionId": sid,
            },
        }).encode("utf-8")

        url = f"{self._config.base_url}/v1/mcp/intercept"
        req = urllib.request.Request(
            url,
            data=payload,
            headers={
                "Content-Type": "application/json",
                "Authorization": f"Bearer {self._config.api_key}",
                "User-Agent": "agentguard-python-sdk/1.0",
            },
            method="POST",
        )

        try:
            with urllib.request.urlopen(req, timeout=self._config.timeout_sec) as resp:
                body = json.loads(resp.read().decode("utf-8"))
                evaluation_ms = (time.monotonic() - start) * 1000
                return McpDecision(
                    allowed=body.get("allowed", False),
                    decision=body.get("decision", "block"),
                    reason=body.get("reason", "Unknown"),
                    risk_score=body.get("riskScore", 0),
                    matched_rule_id=body.get("matchedRuleId"),
                    gate_id=body.get("gateId"),
                    gate_timeout_sec=body.get("gateTimeoutSec"),
                    evaluation_ms=evaluation_ms,
                )
        except Exception as exc:
            evaluation_ms = (time.monotonic() - start) * 1000
            if self._config.strict:
                # Fail closed
                return McpDecision(
                    allowed=False,
                    decision="block",
                    reason=f"AgentGuard evaluation failed (strict mode): {exc}",
                    risk_score=800,
                    evaluation_ms=evaluation_ms,
                )
            else:
                # Fail open
                return McpDecision(
                    allowed=True,
                    decision="allow",
                    reason=f"AgentGuard evaluation failed (permissive mode): {exc}",
                    risk_score=100,
                    evaluation_ms=evaluation_ms,
                )

    async def intercept(
        self,
        tool_name: str,
        arguments: Dict[str, Any],
        session_id: Optional[str] = None,
    ) -> McpDecision:
        """
        Async version of intercept_sync.
        Runs the synchronous HTTP call in a thread pool to avoid blocking.
        """
        loop = asyncio.get_event_loop()
        return await loop.run_in_executor(
            None,
            lambda: self.intercept_sync(tool_name, arguments, session_id),
        )

    def enforce(
        self,
        session_id: Optional[str] = None,
        tool_name_override: Optional[str] = None,
    ) -> Callable:
        """
        Decorator that enforces AgentGuard policy on an async function.

        The decorated function's name is used as the tool name unless
        ``tool_name_override`` is specified.

        Example::

            @guard.enforce(session_id="sess-001")
            async def filesystem_write(path: str, content: str) -> str:
                with open(path, "w") as f:
                    f.write(content)
                return "ok"

        :raises PermissionError: If the tool call is blocked.
        :raises RuntimeError: If HITL approval is pending.
        """
        def decorator(func: Callable) -> Callable:
            tool = tool_name_override or func.__name__

            @wraps(func)
            async def async_wrapper(*args: Any, **kwargs: Any) -> Any:
                # Build arguments dict from call signature
                arguments: Dict[str, Any] = {}
                import inspect
                sig = inspect.signature(func)
                params = list(sig.parameters.keys())
                for i, arg in enumerate(args):
                    if i < len(params):
                        arguments[params[i]] = arg
                arguments.update(kwargs)

                decision = await self.intercept(tool, arguments, session_id)

                if decision.is_blocked:
                    raise PermissionError(
                        f"AgentGuard: Tool '{tool}' is blocked — {decision.reason}"
                    )

                if decision.is_hitl:
                    raise RuntimeError(
                        f"AgentGuard: Tool '{tool}' requires human approval "
                        f"(gateId={decision.gate_id}, timeout={decision.gate_timeout_sec}s)"
                    )

                return await func(*args, **kwargs)

            @wraps(func)
            def sync_wrapper(*args: Any, **kwargs: Any) -> Any:
                # Build arguments dict
                import inspect
                sig = inspect.signature(func)
                params = list(sig.parameters.keys())
                arguments: Dict[str, Any] = {}
                for i, arg in enumerate(args):
                    if i < len(params):
                        arguments[params[i]] = arg
                arguments.update(kwargs)

                decision = self.intercept_sync(tool, arguments, session_id)

                if decision.is_blocked:
                    raise PermissionError(
                        f"AgentGuard: Tool '{tool}' is blocked — {decision.reason}"
                    )

                if decision.is_hitl:
                    raise RuntimeError(
                        f"AgentGuard: Tool '{tool}' requires human approval "
                        f"(gateId={decision.gate_id})"
                    )

                return func(*args, **kwargs)

            if asyncio.iscoroutinefunction(func):
                return async_wrapper
            return sync_wrapper

        return decorator


# ─── wrap_server ──────────────────────────────────────────────────────────────

def wrap_server(
    server: Any,
    api_key: str,
    agent_id: str,
    base_url: str = "https://api.agentguard.tech",
    session_id: Optional[str] = None,
    strict: bool = True,
) -> Any:
    """
    Wrap an MCP server instance with AgentGuard policy enforcement.

    Intercepts the server's tool call handler to evaluate each tool call
    against the agent's policy before execution.

    Works with MCP servers that follow the standard Python MCP SDK pattern:
    - ``server.call_tool(name, arguments)``
    - ``@server.call_tool()`` decorated handlers

    :param server: The MCP server instance to wrap.
    :param api_key: AgentGuard API key.
    :param agent_id: AgentGuard agent ID.
    :param base_url: AgentGuard API base URL.
    :param session_id: Session ID for audit continuity.
    :param strict: Block on policy eval errors (default: True).
    :returns: Wrapped server with the same interface.

    Example::

        from mcp.server import Server
        from agentguard.integrations.mcp import wrap_server

        server = Server("my-mcp-server")
        server = wrap_server(server, api_key="ag_...", agent_id="my-agent")

        @server.call_tool()
        async def filesystem_write(name: str, arguments: dict) -> list:
            path = arguments["path"]
            content = arguments["content"]
            with open(path, "w") as f:
                f.write(content)
            return [{"type": "text", "text": "Written successfully"}]

        # filesystem_write calls now go through AgentGuard policy check
    """
    guard = McpGuard(
        api_key=api_key,
        agent_id=agent_id,
        base_url=base_url,
        session_id=session_id,
        strict=strict,
    )

    # Wrap the server object using a proxy-like pattern
    return _McpServerWrapper(server, guard)


class _McpServerWrapper:
    """
    Internal wrapper that intercepts MCP server tool call handlers.
    """

    def __init__(self, server: Any, guard: McpGuard) -> None:
        self._server = server
        self._guard = guard

    def __getattr__(self, name: str) -> Any:
        attr = getattr(self._server, name)

        # Intercept call_tool decorator/method
        if name == "call_tool":
            return self._wrap_call_tool(attr)

        return attr

    def _wrap_call_tool(self, original: Any) -> Any:
        """
        Wraps the call_tool method/decorator to intercept tool calls.
        """
        if callable(original) and hasattr(original, "__call__"):
            def wrapped_decorator(*args: Any, **kwargs: Any) -> Any:
                decorator = original(*args, **kwargs)

                def inner(handler: Callable) -> Any:
                    @wraps(handler)
                    async def guarded_handler(name: str, arguments: Dict[str, Any], *hargs: Any, **hkwargs: Any) -> Any:
                        decision = await self._guard.intercept(name, arguments or {})

                        if decision.is_blocked:
                            return [{
                                "type": "text",
                                "text": json.dumps({
                                    "error": "blocked",
                                    "reason": decision.reason,
                                    "riskScore": decision.risk_score,
                                }),
                            }]

                        if decision.is_hitl:
                            return [{
                                "type": "text",
                                "text": json.dumps({
                                    "status": "hitl_pending",
                                    "gateId": decision.gate_id,
                                    "gateTimeoutSec": decision.gate_timeout_sec,
                                    "message": f"Tool '{name}' requires human approval",
                                }),
                            }]

                        return await handler(name, arguments, *hargs, **hkwargs)

                    return decorator(guarded_handler) if decorator is not None else guarded_handler

                return inner

            return wrapped_decorator

        return original
